## Упражнения по библиотеке Numpy

In [1]:
from functools import reduce
import numpy as np
import scipy
import random
import time

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [24]:
arr = np.random.randint(10, size = (10,1))
print(arr)
for idx, a in enumerate(arr):
    if((a > 3) & (a < 8)):
        arr[idx] = -a
print(arr)

[[5]
 [4]
 [6]
 [4]
 [4]
 [2]
 [7]
 [5]
 [4]
 [8]]
[[-5]
 [-4]
 [-6]
 [-4]
 [-4]
 [ 2]
 [-7]
 [-5]
 [-4]
 [ 8]]


**2.** Заменить максимальный элемент случайного массива на 0

In [36]:
arr = np.random.randint(10, size = (10,1))
print(arr, "\n max value is", arr.max())
for idx, a in enumerate(arr):
    if(a == arr.max()):
        arr[idx] = 0
print(arr, "new max value is", arr.max())

[[9]
 [0]
 [1]
 [5]
 [6]
 [6]
 [0]
 [8]
 [9]
 [3]] 
 max value is 9
[[0]
 [0]
 [1]
 [5]
 [6]
 [6]
 [0]
 [8]
 [0]
 [3]] new max value is 8


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [37]:
x = np.random.randint(10, size = (2,2))
y = np.random.randint(10, size = (2,2))

print("first array:\n", x, "\nsecond array:\n", y)

cartesian_product = []
for i in range(len(x)):
    for j in range(len(y)):
        cartesian_product.append(str(x[i]) + str(y[j]))

print(cartesian_product)


first array:
 [[6 1]
 [9 0]] 
second array:
 [[9 6]
 [2 0]]
['[6 1][9 6]', '[6 1][2 0]', '[9 0][9 6]', '[9 0][2 0]']


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [47]:
A = np.random.randint(10, size = (8,3))
B = np.random.randint(10, size = (2,2))

print("first array:\n", A, "\nsecond array:\n", B)

b_row1 = B[0,:]
b_row2 = B[1,:]

result = []

for row_A in A:
    first_row = 0
    second_row = 0
    for number in row_A:
        for value_B_row1 in b_row1:
            if number == value_B_row1:
                first_row = 1
        for value_B_row2 in b_row2:
            if number == value_B_row2:
                second_row = 1
    if ((first_row > 0) & (second_row > 0)):
        result.append(row_A)
        
print(result)

first array:
 [[8 9 3]
 [0 7 5]
 [6 6 7]
 [0 4 7]
 [5 3 0]
 [3 7 4]
 [0 2 7]
 [8 1 6]] 
second array:
 [[4 9]
 [7 0]]
[array([0, 4, 7]), array([3, 7, 4])]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [68]:
A = np.array([[2,3,4], [3,3,3], [8,4,5], [4,7,0], [1,1,1], [5,3,1], [7,3,6], [0,0,0], [7,4,3], [8,3,5]])
print(A)

nonuniwue_rows = []
for i, row in enumerate(A):
    if np.unique(row).size == 1:
        nonuniwue_rows.append(i)
        
result = np.delete(A, nonuniwue_rows, axis = 0)
print(result)

[[2 3 4]
 [3 3 3]
 [8 4 5]
 [4 7 0]
 [1 1 1]
 [5 3 1]
 [7 3 6]
 [0 0 0]
 [7 4 3]
 [8 3 5]]
[[2 3 4]
 [8 4 5]
 [4 7 0]
 [5 3 1]
 [7 3 6]
 [7 4 3]
 [8 3 5]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [81]:
A = np.array([[2,3,4], [8,4,5], [8,4,5], [3,3,3], [8,4,5], [5,3,1], [7,3,6], [3,3,3], [7,4,3], [8,3,5]])
print(A)

repeating_lines = []
for i, row1 in enumerate(A):
    for j, row2 in enumerate(A):
        if i != j and np.array_equal(row1, row2) and j not in repeating_lines:
            repeating_lines.append(i)
        
result = np.delete(A, repeating_lines, axis = 0)
print(result)

[[2 3 4]
 [8 4 5]
 [8 4 5]
 [3 3 3]
 [8 4 5]
 [5 3 1]
 [7 3 6]
 [3 3 3]
 [7 4 3]
 [8 3 5]]
[[2 3 4]
 [8 4 5]
 [5 3 1]
 [7 3 6]
 [3 3 3]
 [7 4 3]
 [8 3 5]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [134]:
#np realization
A = np.random.randint(10, size = (4,3))
print("matrix: \n", A)
diagonal = np.diag(A)
print("\ndiagonal: ", diagonal)
filtered = diagonal[diagonal != 0]
result = np.prod(filtered)
print("\nnumpy result: ", result)


#random list to matrix creation 
def create_random_matrix(rows, columns):
    matrix = []
    for i in range(rows):
        row = [random.randint(0, 9) for i in range(columns)]
        matrix.append(row)
    return matrix

#native realization
B = create_random_matrix(5,4)
print("\nmatrix: ", B)
for i in range(len(B[0]), len(B)):
    B.pop()
diagonal = []
for idx, i in enumerate(B):
    if i[idx] != 0:
        diagonal.append(i[idx])
print("\ndiagonal: ", diagonal)
result = reduce((lambda x, y: x * y), diagonal)
print("\nnative result: ", result)

matrix: 
 [[2 8 0]
 [1 3 2]
 [3 6 9]
 [4 5 7]]

diagonal:  [2 3 9]

numpy result:  54

matrix:  [[4, 6, 5, 6], [1, 8, 5, 2], [1, 6, 1, 3], [5, 2, 4, 1], [4, 8, 0, 6]]

diagonal:  [4, 8, 1, 1]

native result:  32


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [55]:
x = np.random.randint(0, 10, 5)
y = np.random.randint(0, 10, 5)
# x = np.array([1,2,2,4])
# y = np.array([4,2,1,2])
print("fist vector is:\n", x, "\nsecond vector is:\n", y)

#numpy realization
if np.array_equal(np.sort(x), np.sort(y)):
    print("vectors define the same multiset")
else:
    print("vectors do not define the same multiset")
    
#native realization
if sorted(x) == sorted(y):
    print("vectors define the same multiset")
else:
    print("vectors do not define the same multiset")

fist vector is:
 [8 3 6 9 4] 
second vector is:
 [8 1 4 2 8]
vectors do not define the same multiset
vectors do not define the same multiset


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [95]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
#x = np.random.randint(0, 10, 9)
print("vector:\n", x)
indices = np.where((x[:-1] == 0) & (x[1:] != 0))[0]
indices = indices + 1
max_element = np.max(x[indices])
print("Maximum element: ", max_element)

#native realization
afternull = []
for i in range(1, len(x)):
    if ((x[i-1] == 0) & (x[i] != 0)):
        afternull.append(x[i])
result = max(afternull)
print("Maximum element: ", result)

vector:
 [6 2 0 3 0 0 5 7 0]
Maximum element:  5
Maximum element:  5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [104]:
#x = np.array([2, 2, 2, 3, 3, 3, 5])
x = np.random.randint(0, 10, 9)
print("vector:\n", x)

#numpy realization
values, counts = np.unique(x, return_counts=True)
print("values in x:\n", values, "\ncounts in x:\n", counts)

#native realization
values = []
counts = []
for i in x:
    entries = 0
    if i not in values:
        values.append(i)
        for j in x:
            if i == j:
                entries = entries +1
        counts.append(entries)

print("values in x:\n", values, "\ncounts in x:\n", counts)

vector:
 [0 9 2 7 6 4 8 4 4]
values in x:
 [0 2 4 6 7 8 9] 
counts in x:
 [1 1 3 1 1 1 1]
values in x:
 [0, 9, 2, 7, 6, 4, 8] 
counts in x:
 [1, 1, 1, 1, 1, 3, 1]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [137]:
x = np.array((1, 2, 3))
y = np.array((4, 5, 6))
print("fist vector is:\n", x, "\nsecond vector is:\n", y)

st = time.time()
euclidean_distance = np.linalg.norm(a - b)
et = time.time()
print("\nnumpy realization:\n", euclidean_distance, "\ntime consumed:\n", et - st)

st = time.time()
euclidean_distance1 = scipy.spatial.distance.euclidean(a, b)
et = time.time()
print("\nscipy realization:\n", euclidean_distance1, "\ntime consumed:\n", et - st)

st = time.time()
sum = 0
for i in range(len(a)):
    sum = sum + (a[i] - b[i])**2
result = sum**(1/2)
et = time.time()
print("\nmanual realization:\n", result, "\ntime consumed:\n", et - st)

fist vector is:
 [1 2 3] 
second vector is:
 [4 5 6]

numpy realization:
 5.196152422706632 
time consumed:
 0.0002219676971435547

scipy realization:
 5.196152422706632 
time consumed:
 5.745887756347656e-05

manual realization:
 5.196152422706632 
time consumed:
 7.82012939453125e-05


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [138]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [2]:
calorie_stats = np.loadtxt("cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [3]:
average_calories = np.average(calorie_stats)
print(average_calories)

106.88311688311688


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [36]:
calorie_stats_sorted = sorted(calorie_stats)
print(calorie_stats_sorted)
print(len(calorie_stats_sorted))

[50.0, 50.0, 50.0, 70.0, 70.0, 80.0, 90.0, 90.0, 90.0, 90.0, 90.0, 90.0, 90.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 100.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 110.0, 120.0, 120.0, 120.0, 120.0, 120.0, 120.0, 120.0, 120.0, 120.0, 120.0, 130.0, 130.0, 140.0, 140.0, 140.0, 150.0, 150.0, 160.0]
77


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [5]:
median_calories = np.median(calorie_stats)
print(median_calories)

110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [32]:
keys = np.arange(3, 4, 0.01)
values = []
dict = {}
for idx, i in enumerate(keys):
    values.append(np.percentile(calorie_stats, i))
    dict[keys[idx]] = values[idx]

print(dict)
for i in keys:
    if dict[i] < 60:
        dict.pop(i)
        
nth_percentile = list(dict.items())[0][0]
print("\nlowest percentile over 60 calories:\n", nth_percentile, np.percentile(calorie_stats, nth_percentile))

{3.0: 55.599999999999994, 3.01: 55.751999999999995, 3.0199999999999996: 55.90399999999999, 3.0299999999999994: 56.05599999999999, 3.039999999999999: 56.20799999999999, 3.049999999999999: 56.359999999999985, 3.0599999999999987: 56.511999999999986, 3.0699999999999985: 56.66399999999997, 3.0799999999999983: 56.815999999999974, 3.089999999999998: 56.96799999999997, 3.099999999999998: 57.11999999999997, 3.1099999999999977: 57.27199999999996, 3.1199999999999974: 57.423999999999964, 3.1299999999999972: 57.57599999999996, 3.139999999999997: 57.72799999999996, 3.149999999999997: 57.87999999999995, 3.1599999999999966: 58.031999999999954, 3.1699999999999964: 58.18399999999994, 3.179999999999996: 58.33599999999994, 3.189999999999996: 58.487999999999936, 3.1999999999999957: 58.63999999999994, 3.2099999999999955: 58.79199999999993, 3.2199999999999953: 58.94399999999992, 3.229999999999995: 59.09599999999993, 3.239999999999995: 59.24799999999992, 3.2499999999999947: 59.39999999999991, 3.25999999999999

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [33]:
more_calories = (np.sum(calorie_stats > 60) / len(calorie_stats)) * 100
print("percentage of producers with calorie counts over 60:\n", more_calories)

percentage of producers with calorie counts over 60:
 96.1038961038961


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [34]:
calorie_std = np.std(calorie_stats)
print(calorie_std)

19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Среднее значение: 
среднее (106) представляет собой среднее арифметическое набора данных, указывая на то, что в среднем количество калорий на порцию составляет около 106.
CrunchieMunchies же содержат 60 каллорий на порцию и могут быть ориентированы на людей, которые следят за своим потреблением калорий или ищут более здоровые варианты хлопьев.

Медиана: 
медиана (110) представляет собой среднее значение набора данных при сортировке в порядке возрастания, указывая на то, что половина данных имеет количество калорий на порцию ниже 110, а половина имеет большее количество калорий, что означает, что CrunchieMunchies предлагают значительное снижение калорий по сравнению с основным количеством доступных для потребителя хлопьев. Можно подчеркнуть потенциальную пользу для здоровья от употребления низкокалорийных хлопьев в составе сбалансированной диеты.

Стандартное отклонение:
стандартное отклонение (19,35718533390827) измеряет дисперсию точек данных вокруг среднего значения. Более высокое значение С.О. указывает на бОльшую изменчивость в наборе данных, что позволит нам выделить и позиционировать CrunchieMunchies как благоприятный выбор для тех, кто ищет определенное количество калорий без ущерба для вкуса или питательных свойств.

Самый низкий процентиль: 
самый низкий процентиль, превышающий 60 калорий, составляет 3,289999999999994, указывая на то, что совсем небольшая часть производителей в нашем наборе данных имеет количество калорий ниже 60.

Процент производителей: 
процент производителей с количеством калорий более 60 составляет 96,1038961038961%.